### PR TITLE
Add new external UCS-related params in PTU, PreloadedPT, SnapshotPT

### PR DESCRIPTION
**Pull request by istoilovgithub**

### CODE REVIEW

**Comment by istoilovgithub** [comment]
Rebuild required

**Comment by istoilovgithub** [comment]
@LevchenkoS @okozlovlux Please review

File: src/components/policy/policy_external/src/policy_table/types.cc [hunk 1]
```diff
@@ -1559,24 +1592,32 @@ ConsentRecords::~ConsentRecords() {}
...
+                       &result__);
   impl::WriteJsonField("input", input, &result__);
   impl::WriteJsonField("time_stamp", time_stamp, &result__);
   return result__;
 }
 
 bool ConsentRecords::is_valid() const {
   if (struct_empty()) {
-    return initialization_state__ == kInitialized && Validate();
+    return initialization_state__ == kUninitialized && Validate();
```
Comment (LevchenkoS) [inline src/components/policy/policy_external/src/policy_table/types.cc]: @istoilovgithub, are you sure about this change?

Comment (istoilovgithub) [inline src/components/policy/policy_external/src/policy_table/types.cc]: @LevchenkoS Honestly my initial reaction to this change was like yours, since this change is merged and slightly modified from already implemented functionality lets ask it's initial author to give you a thorough explanation, which I am currently is unable to provide in deep. @dev-gh, please comment.

Comment (dev-gh) [inline src/components/policy/policy_external/src/policy_table/types.cc]: @istoilovgithub 
I suppose that was an error, please revert this line.

Comment (istoilovgithub) [inline src/components/policy/policy_external/src/policy_table/types.cc]: @LevchenkoS fixed with 6145f14

File: src/components/policy/policy_external/include/policy/policy_table/types.h [hunk 2]
```diff
@@ -229,10 +229,35 @@ struct RpcParameters : CompositeType {
...
+struct ExternalConsentEntity : CompositeType {
+ public:
+  typedef Integer<int32_t, INT32_MIN, INT32_MAX> EntityInt;
+  EntityInt entity_type;
+  EntityInt entity_id;
+
+ public:
+  ExternalConsentEntity();
+  explicit ExternalConsentEntity(const Json::Value* value__);
+  ExternalConsentEntity(int32_t type, int32_t id);
```
Comment (istoilovgithub) [inline src/components/policy/policy_external/include/policy/policy_table/types.h]: @okozlovlux fixed with 6145f14

File: src/components/policy/policy_external/test/sql_pt_representation_test.cc [hunk 3]
```diff
@@ -62,10 +62,15 @@ using testing::ReturnRef;
 using testing::Return;
 using testing::NiceMock;
 using testing::Mock;
+
 namespace test {
 namespace components {
 namespace policy_test {
 
+namespace {
+const int32_t policy_tables_number = 27;
```
Comment (okozlovlux) [inline src/components/policy/policy_external/test/sql_pt_representation_test.cc]: @istoilovgithub , this const must be named `kPolicyTablesNumber`

Comment (istoilovgithub) [inline src/components/policy/policy_external/test/sql_pt_representation_test.cc]: @okozlovlux fixed with e35e76c

File: src/components/policy/policy_external/include/policy/policy_table/types.h [hunk 4]
```diff
@@ -229,10 +229,35 @@ struct RpcParameters : CompositeType {
...
+  EntityInt entity_type;
+  EntityInt entity_id;
+
+ public:
+  ExternalConsentEntity();
+  explicit ExternalConsentEntity(const Json::Value* value__);
+  ExternalConsentEntity(const int32_t type, const int32_t id);
+  Json::Value ToJsonValue() const;
+  bool operator==(const ExternalConsentEntity& rhs) const;
+  bool is_valid() const;
```
Comment (LuxoftAKutsan) [inline src/components/policy/policy_external/include/policy/policy_table/types.h]: @istoilovgithub please mark appropriate methods as OVERRIDE.

File: src/components/policy/policy_external/src/policy_table/validation.cc [hunk 5]
```diff
@@ -206,6 +206,10 @@ bool UsageAndErrorCounts::Validate() const {
   return true;
 }
 bool ConsentRecords::Validate() const {
+  if (PT_SNAPSHOT != GetPolicyTableType()) {
+    return !external_consent_status_groups->is_initialized();
+  }
+
```
Comment (LuxoftAKutsan) [inline src/components/policy/policy_external/src/policy_table/validation.cc]: @istoilovgithub is external_consent_status_groups is allowed for PRELOADED or UPDATE?

Comment (LuxoftAKutsan) [inline src/components/policy/policy_external/src/policy_table/validation.cc]: @istoilovgithub should be any changes in Rpcs::Validate?

File: src/components/policy/policy_external/src/sql_pt_representation.cc [hunk 6]
```diff
@@ -616,19 +616,32 @@ bool SQLPTRepresentation::GatherFunctionalGroupings(
...
+
     if (!func_group.IsNull(2)) {
       *rpcs_tbl.user_consent_prompt = func_group.GetString(2);
     }
-    int func_id = func_group.GetInteger(0);
-    rpcs.Bind(0, func_id);
+
+    const int group_id = func_group.GetInteger(0);
+
+    rpcs.Bind(0, group_id);
```
Comment (LuxoftAKutsan) [inline src/components/policy/policy_external/src/sql_pt_representation.cc]: @istoilovgithub why did you add this spaces?

File: src/components/policy/policy_external/src/sql_pt_representation.cc [hunk 7]
```diff
@@ -649,10 +662,29 @@ bool SQLPTRepresentation::GatherFunctionalGroupings(
         }
       }
     }
+
+    rpcs.Reset();
```
Comment (LuxoftAKutsan) [inline src/components/policy/policy_external/src/sql_pt_representation.cc]: @istoilovgithub why did you moved Reset?

File: src/components/policy/policy_external/src/sql_pt_representation.cc [hunk 8]
```diff
@@ -649,10 +662,29 @@ bool SQLPTRepresentation::GatherFunctionalGroupings(
...
+    // Collecting entities for disallowed_by_external_consent_entities_on/off
+    external_consent_entities.Bind(0, group_id);
+    while (external_consent_entities.Next()) {
+      policy_table::ExternalConsentEntity external_consent_entity(
+          external_consent_entities.GetInteger(0),
+          external_consent_entities.GetInteger(1));
+
+      policy_table::DisallowedByExternalConsentEntities&
+          external_consent_entities_container =
+              "ON" == external_consent_entities.GetString(2)
```
Comment (LuxoftAKutsan) [inline src/components/policy/policy_external/src/sql_pt_representation.cc]: @istoilovgithub can you void such magic? At lease make local const string with name.

File: src/components/policy/policy_external/src/sql_pt_representation.cc [hunk 9]
```diff
@@ -649,10 +662,29 @@ bool SQLPTRepresentation::GatherFunctionalGroupings(
...
+                  ? *rpcs_tbl.disallowed_by_external_consent_entities_off
+                  : *rpcs_tbl.disallowed_by_external_consent_entities_off;
```
Comment (LuxoftAKutsan) [inline src/components/policy/policy_external/src/sql_pt_representation.cc]: @istoilovgithub typo? I suppose one of them should be on. Please mention to create something like UT to avoid such typos.

File: src/components/policy/policy_external/src/sql_pt_representation.cc [hunk 10]
```diff
@@ -804,10 +842,27 @@ bool SQLPTRepresentation::SaveFunctionalGroupings(
       return false;
     }
 
-    if (!SaveRpcs(query.LastInsertId(), it->second.rpcs)) {
+    const int64_t last_group_id = query.LastInsertId();
+
+    if (!SaveRpcs(last_group_id, it->second.rpcs)) {
```
Comment (LuxoftAKutsan) [inline src/components/policy/policy_external/src/sql_pt_representation.cc]: @istoilovgithub unable to find what is ```it```. Please  name it apropriatary

File: src/components/policy/policy_external/src/sql_pt_representation.cc [hunk 11]
```diff
@@ -869,6 +924,37 @@ bool SQLPTRepresentation::SaveRpcs(int64_t group_id,
...
+    const policy_table::DisallowedByExternalConsentEntities& entities,
+    ExternalConsentEntitiesType type) const {
+  utils::dbms::SQLQuery query(db());
+  if (!query.Prepare(sql_pt::kInsertExternalConsentEntity)) {
+    LOG4CXX_WARN(logger_,
+                 "Incorrect insert statement for external consent entities.");
+    return false;
+  }
+
+  const std::string on_off =
```
Comment (LuxoftAKutsan) [inline src/components/policy/policy_external/src/sql_pt_representation.cc]: @istoilovgithub on_off is not the best name. probably entities_type_status, external_concents_status.

File: src/components/policy/policy_external/test/sql_pt_representation_test.cc [hunk 12]
```diff
@@ -62,10 +62,15 @@ using testing::ReturnRef;
 using testing::Return;
 using testing::NiceMock;
 using testing::Mock;
+
 namespace test {
 namespace components {
 namespace policy_test {
 
+namespace {
+const int32_t kPolicyTablesNumber = 27;
```
Comment (LuxoftAKutsan) [inline src/components/policy/policy_external/test/sql_pt_representation_test.cc]: @istoilovgithub what is it? Why 27?

File: src/components/policy/policy_external/test/sql_pt_representation_test.cc [hunk 13]
```diff
@@ -401,7 +406,7 @@ TEST_F(SQLPTRepresentationTest,
   // Check PT structure destroyed and tables number is 0
   query.Prepare(query_select);
   query.Next();
-  ASSERT_EQ(25, query.GetInteger(0));
+  ASSERT_EQ(kPolicyTablesNumber, query.GetInteger(0));
```
Comment (LuxoftAKutsan) [inline src/components/policy/policy_external/test/sql_pt_representation_test.cc]: @dev-gh I can see that you asser reading this constant. But I can't see that you insert it somewhere.

File: src/components/policy/policy_external/src/sql_pt_representation.cc (hunk 6 above)
Comment (istoilovgithub) [inline src/components/policy/policy_external/src/sql_pt_representation.cc]: @LuxoftAKutsan More or less it is a subjective matter and it is hard to be given the right answer, right?

File: src/components/policy/policy_external/src/sql_pt_representation.cc (hunk 7 above)
Comment (istoilovgithub) [inline src/components/policy/policy_external/src/sql_pt_representation.cc]: @LuxoftAKutsan Looks better to me, that was the reason. Probably more important question would be: does this change make any functional difference. Looks like not, could you please confirm.

File: src/components/policy/policy_external/src/policy_table/validation.cc (hunk 5 above)
Comment (istoilovgithub) [inline src/components/policy/policy_external/src/policy_table/validation.cc]: @LuxoftAKutsan  According to requirements external_consent_status_groups is allowed to be present only in snapshots. The code already has contained required validation, see ConsentRecords::is_valid.

File: src/components/policy/policy_external/include/policy/policy_table/types.h (hunk 4 above)
Comment (istoilovgithub) [inline src/components/policy/policy_external/include/policy/policy_table/types.h]: @LuxoftAKutsan It has been done already, just pay attention to the few more lines below.

File: src/components/policy/policy_external/src/sql_pt_representation.cc (hunk 8 above)
Comment (istoilovgithub) [inline src/components/policy/policy_external/src/sql_pt_representation.cc]: @LuxoftAKutsan Fixed with e9504ab

File: src/components/policy/policy_external/src/sql_pt_representation.cc (hunk 9 above)
Comment (istoilovgithub) [inline src/components/policy/policy_external/src/sql_pt_representation.cc]: @LuxoftAKutsan Fixed with e9504ab

File: src/components/policy/policy_external/src/sql_pt_representation.cc (hunk 10 above)
Comment (istoilovgithub) [inline src/components/policy/policy_external/src/sql_pt_representation.cc]: @LuxoftAKutsan Fixed with e9504ab

File: src/components/policy/policy_external/src/sql_pt_representation.cc (hunk 11 above)
Comment (istoilovgithub) [inline src/components/policy/policy_external/src/sql_pt_representation.cc]: @LuxoftAKutsan Fixed with e9504ab

File: src/components/policy/policy_external/test/sql_pt_representation_test.cc (hunk 12 above)
Comment (istoilovgithub) [inline src/components/policy/policy_external/test/sql_pt_representation_test.cc]: @LuxoftAKutsan The answer for your question could be found by taking a look at the related change in RefreshDB_DropExistedPTThenRefreshDB_ExpectTablesWithInitialData test.

File: src/components/policy/policy_external/include/policy/policy_table/types.h (hunk 4 above)
Comment (LuxoftAKutsan) [inline src/components/policy/policy_external/include/policy/policy_table/types.h]: @istoilovgithub is_valid is virtual method as I remember same as  is_initialized and ReportErrors . Makr them as OVERRIDE

File: src/components/policy/policy_external/test/sql_pt_representation_test.cc (hunk 13 above)
Comment (LuxoftAKutsan) [inline src/components/policy/policy_external/test/sql_pt_representation_test.cc]: @istoilovgithub what about this issue?

File: src/components/policy/policy_external/include/policy/policy_table/types.h (hunk 4 above)
Comment (istoilovgithub) [inline src/components/policy/policy_external/include/policy/policy_table/types.h]: @LuxoftAKutsan Correct me if I am wrong, but quick reference to the base class in [rpc_base.h](sdl_core/src/components/rpc_base/include/rpc_base/rpc_base.h) shows that all applicable methods have been marked already.

File: src/components/policy/policy_external/test/sql_pt_representation_test.cc (hunk 13 above)
Comment (istoilovgithub) [inline src/components/policy/policy_external/test/sql_pt_representation_test.cc]: @LuxoftAKutsan This magic number comes from the kCreateSchema query string, see [sql_pt_queries.cc](https://github.com/smartdevicelink/sdl_core/blob/feature/extended_policy_functionality/src/components/policy/policy_regular/src/sql_pt_queries.cc). Before changes tables number was 25 (just search for 'CREATE TABLE') and after changes: +2.

**Comment by LuxoftAKutsan** [comment]
:+1:
